contracts/0.8.9/LidoLocator.sol

```diff
@@ -0,0 +1,79 @@
+// SPDX-FileCopyrightText: 2023 Lido <info@lido.fi>
+
+// SPDX-License-Identifier: GPL-3.0
+
+/* See contracts/COMPILERS.md */
+pragma solidity 0.8.9;
+
+import {ILidoLocator} from "../common/interfaces/ILidoLocator.sol";
+
+/**
+ * @title LidoLocator
+ * @author mymphe
+ * @notice Lido service locator
+ * @dev configuration is stored as public immutables to reduce gas consumption
+ */
+contract LidoLocator is ILidoLocator {
+    struct Config {
+        address accountingOracle;
+        address depositSecurityModule;
+        address elRewardsVault;
+        address legacyOracle;
+        address lido;
+        address safetyNetsRegistry;
+        address selfOwnedStEthBurner;
+        address stakingRouter;
+        address treasury;
+        address validatorExitBus;
+        address withdrawalQueue;
+        address withdrawalVault;
+    }
+
+    error ErrorZeroAddress();
+
+    address public immutable accountingOracle;
+    address public immutable depositSecurityModule;
+    address public immutable elRewardsVault;
+    address public immutable legacyOracle;
+    address public immutable lido;
+    address public immutable safetyNetsRegistry;
+    address public immutable selfOwnedStEthBurner;
+    address public immutable stakingRouter;
+    address public immutable treasury;
+    address public immutable validatorExitBus;
+    address public immutable withdrawalQueue;
+    address public immutable withdrawalVault;
+
+    /**
+     * @notice declare service locations
+     * @dev accepts a struct to avoid the "stack-too-deep" error
+     * @param _config struct of addresses
+     */
+    constructor(Config memory _config) {
```

Comment:
`calldata` would be cheaper

Reply:
Btw, all checks below can be replaced by a universal one that doesn't depend on the number of addresses in the struct:

```solidity
constructor(Config calldata _config) {
    if (!_checkCalldataNonZero()) revert ErrorZeroAddress();
    // make all assignments
}

function _checkCalldataNonZero() internal pure returns (bool allNonZero) {
    assembly {
        allNonZero := 1
        // skip selector
        let offset := 4
        let maxOffset := calldatasize()
        for {} lt(offset, maxOffset) {} {
            if eq(calldataload(offset), 0) {
                allNonZero := 0
                break
            }
            offset := add(offset, 32)
        }
    }
}
```

Reply:
Ah, my bad, it's a constructor :(

Reply:
No `calldata` here unfortunately

Reply:
Btw, how about this?

```solidity
constructor(Config memory _config) {
    accountingOracle = _assertNonZero(_config.accountingOracle);
    // ...
}

function _assertNonZero(address _addr) internal pure {
    if (_addr == address(0)) revert ErrorZeroAddress();
}
```

It would still avoid code duplication and the possibility of forgotten checks.

Reply:
Good idea. Thank you!